python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,572 @@
+        results.append(result)
+
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, "auto")
+
+    x, y = generate_x_y(CONFIG)
+    x = x.to(device)
+    y = y.to(device)
+    model = Model(CONFIG.layer_size, CONFIG.num_layers, device, CONFIG.dtype).to(device)
+    for i in range(0, len(model.layers), 2):
+        model.layers[i].weight = nn.Parameter(results[-1][i // 2])
+    pred = model(y)
+    loss = model.criterion(pred, y)
+    print(f"auto pred: {pred}")
+    print(f"auto loss: {loss}")
+
+    return results
+
+
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+    os.environ["MASTER_PORT"] = "12355"
+
+    # initialize the process group
+    dist.init_process_group("gloo", rank=rank, world_size=world_size)
+
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+def demo_basic(rank, world_size, auto_res, ray_res):
+    print(f"Running basic DDP example on rank {rank}.")
+    setup(rank, world_size)
+
+    # create model and move it to GPU with id rank
+    model = Model(
+        CONFIG.layer_size, CONFIG.num_layers, f"cuda:{rank}", CONFIG.dtype
+    ).to(rank)
+    ddp_model = DDP(model, device_ids=[rank])
+
+    loss_fn = model.criterion
+    optimizer = optim.SGD(ddp_model.parameters(), lr=model.lr)
+
+    num_actors = CONFIG.num_actors
+
+    results = []
+    elapses = []
+    for i in range(CONFIG.it):
+        x, y = generate_x_y(CONFIG)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        x = xs[rank]
+        y = ys[rank]
+        x = x.to(rank)
+        y = y.to(rank)
+        start = time.perf_counter()
+        optimizer.zero_grad()
+        outputs = ddp_model(x)
+        labels = y
+        loss = loss_fn(outputs, labels)
+        loss.backward()
+        optimizer.step()
+        end = time.perf_counter()
+
+        result = []
+        for i in range(0, len(model.layers), 2):
+            layer: torch.nn.Linear = model.layers[i]
+            result.append(torch.clone(layer.weight))
+        results.append(result)
+
+        elapse = end - start
+        elapses.append(elapse)
+
+    print_elapses(elapses, f"dist #{rank}")
+
+    x, y = generate_x_y(CONFIG)
+    x = x.to(rank)
+    y = y.to(rank)
+    model = Model(
+        CONFIG.layer_size, CONFIG.num_layers, f"cuda:{rank}", CONFIG.dtype
+    ).to(rank)
+    for i in range(0, len(model.layers), 2):
+        model.layers[i].weight = nn.Parameter(results[-1][i // 2])
+    pred = model(y)
+    loss = loss_fn(pred, y)
+    print(f"dist #{rank} pred: {pred}")
+    print(f"dist #{rank} loss: {loss}")
+
+    cleanup()
+    compare_results(auto_res, results, "auto vs dist")
+    compare_results(ray_res, results, "ray vs dist")
+
+
+def run_demo(demo_fn, world_size, auto_res, ray_res):
+    mp.spawn(
+        demo_fn, args=(world_size, auto_res, ray_res), nprocs=world_size, join=True
+    )
+
+
+def check_torch_model_correctness_dist(auto_res, ray_res):
```

Comment:
Can you put similar functions like `check_*` in sequence?

Reply:
Done